hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/token/impl/CryptoPreTransactionHandlerImpl.java

```diff
@@ -102,4 +119,49 @@ private TransactionMetadata createAccountSigningMetadata(
         }
         return new SigTransactionMetadata(accountStore, tx, payer);
     }
+
+    /**
+     * Fetches given accountId's key and add it to the metadata only if the accountId is not same as
+     * payer. If the accountId could not be fetched successfully, sets the failure status on the
+     * metadata.
+     *
+     * @param accountId given accountId
+     * @param payer payer accountId
+     * @param meta metadata to which accountId's key will be added, if success
+     */
+    private void addIfNotPayer(
+            final AccountID accountId, final AccountID payer, final SigTransactionMetadata meta) {
+        if (meta.failed() || payer.equals(accountId)) {
+            return;
+        }
+        var result = accountStore.getKey(accountId);
+        if (result.failed()) {
+            meta.setStatus(result.failureReason());
+        } else {
+            meta.addToReqKeys(result.key());
+        }
+    }
+
+    /**
+     * Fetches given accountId's key and add it to the metadata requiredKeys list only if the
+     * accountId is not same as payer and the account has receiverSigRequired flag set to true. If
+     * the accountId have receiverSigRequired flag set false, the key will not be added metadata. If
+     * the accountId could not be fetched successfully, sets the failure status on the metadata.
+     *
+     * @param accountId given accountId
+     * @param payer payer accountId
+     * @param meta metadata to which accountId's key will be added to the requiredKeys, if success
+     */
+    private void addIfNotPayerAndReceiverSigRequired(
+            final AccountID accountId, final AccountID payer, final SigTransactionMetadata meta) {
+        if (meta.failed() || payer.equals(accountId)) {
+            return;
+        }
+        var result = accountStore.getKeyIfReceiverSigRequired(accountId);
+        if (result.failed()) {
+            meta.setStatus(result.failureReason());
+        } else if (result.key() != null) {
```

Comment:
I'd rather put this logic in the `result`, like:
```
    public record KeyOrLookupFailureReason(
            @Nullable HederaKey key, @Nullable ResponseCodeEnum failureReason) {
        public static final KeyOrLookupFailureReason PRESENT_BUT_NOT_REQUIRED = new KeyOrLookupFailureReason(null, null);

        public void addIfPresentAndRequired(final SigTransactionMetadata meta) {
            if (failureReason != null) {
                throw new IllegalStateException("A failed lookup cannot be trusted to add its key to sig meta");
            }
            // If we've already failed in computing earlier sig reqs, do nothing; otherwise
            // add our key to the list, if it's required
            if (!meta.failed() && key != null) {
                meta.addToReqKeys(result.key());
            }
        }
    }
```

Notice this also gives us the option to drop a bunch of explicit `if (!meta.failed() ...` checks like above on L134.

Reply:
Agreed. Thanks for the suggestions. Surely makes it cleaner.